### PR TITLE
view change revamp - 1

### PR DESCRIPTION
**Pull request by LeoHChen**
This is the first view change revamp PR.

It encapsulate all view change related variables and functions into view_change.go, view_change_service.go.

It rewrites the onViewChange function to be more streamline and understandable.

It also add smaller granularity to protect the view change data structures.

### CODE REVIEW

File: consensus/quorum/quorum.go [hunk 1]
```diff
@@ -124,6 +125,7 @@ type Decider interface {
 	IsQuorumAchieved(Phase) bool
 	IsQuorumAchievedByMask(mask *bls_cosi.Mask) bool
 	QuorumThreshold() numeric.Dec
+	OneThirdQuorum() numeric.Dec
```
Comment (rlan35) [inline consensus/quorum/quorum.go:128]: can remove this interface if we won't use it to avoid confusion.

File: consensus/view_change.go [hunk 2]
```diff
@@ -1,687 +1,350 @@
...
-}
+// ViewChangeMsgType defines the type of view change message
+type ViewChangeMsgType byte
+
+const (
+	// M1 is the view change message with block
+	M1 ViewChangeMsgType = iota
+	// M2 is the view change message w/o block
+	M2
+	// M3 is the view change message with viewID ??
```
Comment (rlan35) [inline consensus/view_change.go:29]: M3 Is the signature signed on viewID.

File: consensus/view_change_service.go [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+	pm.viewMux.RLock()
+	pm.cViewMux.RLock()
+	defer pm.viewMux.RUnlock()
+	defer pm.cViewMux.RUnlock()
+	diff := int64(pm.viewChangingID - pm.blockViewID)
+	return time.Duration(diff * diff * int64(viewChangeDuration))
+}
+
+// switchPhase will switch FBFTPhase to nextPhase if the desirePhase equals the nextPhase
+func (consensus *Consensus) switchPhase(subject string, desired FBFTPhase, override bool) {
```
Comment (rlan35) [inline consensus/view_change_service.go:102]: This should be in a general consensus file, not in view_change_service.

File: consensus/view_change_service.go [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+		}
+		msgToSend := consensus.constructViewChangeMessage(&key)
+		consensus.host.SendMessageToGroups([]nodeconfig.GroupID{
+			nodeconfig.NewGroupIDByShardID(nodeconfig.ShardID(consensus.ShardID)),
+		},
+			p2p.ConstructMessage(msgToSend),
+		)
+	}
+
+	consensus.consensusTimeout[timeoutViewChange].SetDuration(duration)
```
Comment (rlan35) [inline consensus/view_change_service.go:182]: duplicate codes.

File: consensus/view_change.go [hunk 5]
```diff
@@ -1,687 +1,350 @@
...
 			}
+			vc.nilSigs[viewID][key.Pub.Bytes.Hex()] = key.Pri.SignHash(NIL)
 		}
-
-		consensus.SetViewChangingID(recvMsg.ViewID)
-		msgToSend := consensus.constructNewViewMessage(
-			recvMsg.ViewID, newLeaderPriKey,
-		)
-
-		consensus.getLogger().Warn().
```
Comment (rlan35) [inline consensus/view_change.go:483]: Let's keep this message? I search for this message for view change debug a lot. You can check it to Info()

File: consensus/quorum/quorum.go (hunk 1 above)
Comment (LeoHChen) [reply consensus/quorum/quorum.go:128]: yes

File: consensus/view_change_service.go (hunk 3 above)
Comment (LeoHChen) [reply consensus/view_change_service.go:102]: will move it.

File: consensus/view_change.go (hunk 2 above)
Comment (LeoHChen) [reply consensus/view_change.go:29]: will remove it,

File: consensus/view_change_service.go (hunk 4 above)
Comment (LeoHChen) [reply consensus/view_change_service.go:182]: will remove.

File: consensus/view_change.go (hunk 5 above)
Comment (LeoHChen) [reply consensus/view_change.go:483]: sure.

**Review by LeoHChen — COMMENTED** [review]
Thanks for the review comments. I've started a new PR. https://github.com/harmony-one/harmony/pull/3373

**Comment by LeoHChen** [comment]
A new PR is started, will address all the comments in the new PR.

https://github.com/harmony-one/harmony/pull/3373

**Pull request closed without merging**